app/scripts/modules/core/src/notification/modal/NotificationDetails.tsx

```diff
@@ -0,0 +1,120 @@
+  constructor(props: INotificationDetailsProps) {
+    super(props);
+
+    let whenOptions = [];
+    if (props.level === 'application' || props.level === 'pipeline') {
+      whenOptions = PIPELINE_WHEN_OPTIONS;
+    } else if (props.stageType === 'manualJudgment') {
+      whenOptions = MANUAL_JUDGEMENT_WHEN_OPTIONS;
+    } else {
+      whenOptions = STAGE_WHEN_OPTIONS;
+    }
+
+    this.state = {
+      notificationTypes: Registry.pipeline.getNotificationTypes().map((type: INotificationTypeConfig) => ({
+        label: type.label,
+        value: type.key,
+      })),
+      whenOptions,
+    };
+  }
+
+  public componentDidMount() {
+    const { formik } = this.props;
+    if (!formik.values.type) {
+      const { notificationTypes } = this.state;
+      formik.setFieldValue('type', notificationTypes ? notificationTypes[0].value : '');
+    }
+  }
+
+  private renderCustomMessage = (type: string, whenOption: string, selectedOptions: string[]): React.ReactNode => {
+    if (
+      whenOption !== 'manualJudgment' &&
+      ['email', 'slack', 'googlechat'].includes(type) &&
+      selectedOptions.includes(whenOption)
+    ) {
+      return (
+        <FormikFormField
+          name={'message.' + whenOption + '.text'}
+          input={props => <TextAreaInput {...props} placeholder="enter a custom notification message (optional)" />}
+        />
+      );
+    } else {
+      return <></>;
+    }
+  };
+
+  private onNotificationTypeChange = (type: string) => {
+    const notificationTypeUpdate = Registry.pipeline.getNotificationConfig(type);
+    if (!!notificationTypeUpdate) {
+      this.props.formik.setFieldValue('address', '');
+    }
+    this.props.formik.setFieldValue('when', [...this.props.formik.values.when]);
+  };
+
+  public validate(_values: INotification) {
+    const { when } = this.props.formik.values;
+    const errors = {} as any;
+    if (!when || when.length === 0) {
+      errors.whenOption = 'Please select when the notification should execute';
+    }
+
+    return errors;
+  }
+
+  public render() {
+    const { onNotificationTypeChange, renderCustomMessage } = this;
+    const { formik, level, stageType } = this.props;
+    const { values } = formik;
+    const { whenOptions } = this.state;
+    return (
+      <>
+        <NotificationSelector onNotificationTypeChange={onNotificationTypeChange} type={values.type} />
+        {(stageType || level) && (
+          <div className="form-group row">
+            <FormikFormField
+              name="when"
+              label="Notify when"
+              input={props => (
+                <ChecklistInput
+                  {...props}
+                  options={whenOptions.map(o => ({
+                    value: o,
+                    label: NotificationTransformer.getNotificationWhenDisplayName(o, level, stageType),
+                    additionalFields: renderCustomMessage(values.type, o, formik.values.when),
```

Comment:
I'm not sure how I feel about this addition to the API.  @alanmquach care to take a look?
![Screen Shot 2019-07-18 at 4 45 05 PM](https://user-images.githubusercontent.com/2053478/61499267-79243b80-a97b-11e9-9f54-96d1238e3cbf.png)

Reply:
I feel like this addition may be overloading the `ChecklistInput` to do something pretty specific.

It would probably be better to just compose together 3 checkboxes and 3 textareas instead of trying to have all this handled completely within `ChecklistInput`

Reply:
I was fighting back and forth with that idea. because I was going to have to reuse a lot of the logic.

Reply:
changed `ChecklistInput` back and implement a component to handle the above case.